js/style/style_declaration.js

```diff
@@ -23,20 +23,32 @@ function StyleDeclaration(propType, renderType, prop, value) {
 
 }
 
-StyleDeclaration.prototype.calculate = function(z) {
-    return typeof this.value === 'function' ? this.value(z) : this.value;
+StyleDeclaration.prototype.calculate = function(z, zoomHistory) {
+    return typeof this.value === 'function' ? this.value(z, zoomHistory) : this.value;
 };
 
-StyleDeclaration.prototype.parseValue = function(value, type, values) {
+StyleDeclaration.prototype.parseValue = function(value, type, values, prop) {
     return type === 'color' ? parseColor(value) :
         type === 'number' ? parseNumber(value) :
         type === 'boolean' ? Boolean(value) :
         type === 'image' ? String(value) :
+        type === 'string' && (prop === 'fill-image' || prop === 'line-image' || prop === 'background-image') ? parsePattern(value) :
```

Comment:
Can we do this in a way that's not hard-coded to particular property names? Either support non-interpolated functions for all properties, or look to the style spec to tell whether functions are supported for a particular property.